go/vt/tabletserver/dbconn.go

```diff
@@ -58,17 +59,21 @@ func (dbc *DBConn) Exec(ctx context.Context, query string, maxrows int, wantfiel
 		case err == nil:
 			return r, nil
 		case !IsConnErr(err):
-			return nil, NewTabletErrorSql(ErrFail, err)
+			// MySQL error that isn't due to a connection issue
+			return nil, NewTabletErrorSql(ErrFail, vtrpc.ErrorCode_UNKNOWN_ERROR, err)
 		case attempt == 2:
-			return nil, NewTabletErrorSql(ErrFatal, err)
+			// If the MySQL connection is bad, we assume that there is nothing wrong with
+			// the query itself, and retrying it might succeed. The MySQL connection might
+			// fix itself, or the query could succeed on a different VtTablet.
+			return nil, NewTabletErrorSql(ErrFatal, vtrpc.ErrorCode_INTERNAL_ERROR, err)
 		}
 		err2 := dbc.reconnect()
 		if err2 != nil {
 			go checkMySQL()
-			return nil, NewTabletErrorSql(ErrFatal, err)
+			return nil, NewTabletErrorSql(ErrFatal, vtrpc.ErrorCode_INTERNAL_ERROR, err)
 		}
 	}
-	return nil, NewTabletErrorSql(ErrFatal, errors.New("dbconn.Exec: unreachable code"))
+	return nil, NewTabletErrorSql(ErrFatal, vtrpc.ErrorCode_INTERNAL_ERROR, errors.New("dbconn.Exec: unreachable code"))
```

Comment:
I know the compiler catches some unreachable code paths, but it seems like it doesn't work for everything. It doesn't catch this case.

I'm going to leave this code as-is.